File: backend/app/schemas/documents.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DocumentOwnerResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    username: str
    nickname: str


class DocumentResponse(BaseModel):
    id: UUID
    group_id: UUID | None
    title: str
    summary: str | None
    category: str | None
    file_name: str
    file_type: str
    mime_type: str
    file_extension: str
    file_size: int
    visibility_mode: str
    status: str
    preview_status: str
    allow_download: bool
    read_count: int
    like_count: int
    coin_count: int
    download_count: int
    password_enabled: bool
    my_liked: bool
    owner: DocumentOwnerResponse
    created_at: datetime
    updated_at: datetime


class DocumentDetailResponse(DocumentResponse):
    specific_usernames: list[str]
    latest_storage_key: str
    inline_preview_supported: bool
    preview_text_available: bool
    preview_strategy: str


class DocumentListResponse(BaseModel):
    items: list[DocumentResponse]


class DocumentLikeResponse(BaseModel):
    liked: bool
    like_count: int


class DocumentCoinRequest(BaseModel):
    coin_amount: int = Field(ge=1, le=100)


class DocumentCoinResponse(BaseModel):
    coin_amount: int
    my_balance: int
    document_coin_count: int
    owner_balance: int
    message: str

File: frontend/lib/documents.ts
import { VisibilityMode, visibilityOptions } from "@/lib/groups";

export type DocumentOwner = {
  id: string;
  username: string;
  nickname: string;
};

export type DocumentSummary = {
  id: string;
  group_id: string | null;
  title: string;
  summary: string | null;
  category: string | null;
  file_name: string;
  file_type: string;
  mime_type: string;
  file_extension: string;
  file_size: number;
  visibility_mode: VisibilityMode;
  status: string;
  preview_status: string;
  allow_download: boolean;
  read_count: number;
  like_count: number;
  coin_count: number;
  download_count: number;
  password_enabled: boolean;
  my_liked: boolean;
  owner: DocumentOwner;
  created_at: string;
  updated_at: string;
};

export type DocumentDetail = DocumentSummary & {
  specific_usernames: string[];
  latest_storage_key: string;
  inline_preview_supported: boolean;
  preview_text_available: boolean;
  preview_strategy: "browser_inline" | "text" | "download_only";
};

export type DocumentListResponse = {
  items: DocumentSummary[];
};

export type DocumentCoinResponse = {
  coin_amount: number;
  my_balance: number;
  document_coin_count: number;
  owner_balance: number;
  message: string;
};

export const documentVisibilityOptions = visibilityOptions;

export const previewStatusLabelMap: Record<string, string> = {
  pending: "待处理",
  processing: "处理中",
  ready: "可预览",
  failed: "处理失败",
};

export function formatFileSize(fileSize: number): string {
  if (fileSize < 1024) {
    return `${fileSize} B`;
  }
  if (fileSize < 1024 * 1024) {
    return `${(fileSize / 1024).toFixed(1)} KB`;
  }
  return `${(fileSize / 1024 / 1024).toFixed(2)} MB`;
}
